cpp/src/arrow/compute/exec/aggregate_node.cc

```diff
@@ -326,46 +446,86 @@ class ScalarAggregateNode : public ExecNode, public TracedNode {
   }
 
  private:
-  Status Finish() {
-    auto scope = TraceFinish();
+  Status ReconstructAggregates() {
+    const auto& input_schema = *inputs()[0]->output_schema();
+    auto exec_ctx = plan()->query_context()->exec_context();
+    for (size_t i = 0; i < kernels_.size(); ++i) {
+      std::vector<TypeHolder> in_types;
+      for (const auto& target : target_fieldsets_[i]) {
+        in_types.emplace_back(input_schema.field(target)->type().get());
+      }
+      states_[i].resize(plan()->query_context()->max_concurrency());
```

Comment:
Why is state resized to max_concurrency here?

Reply:
I don't imagine we would need to resize number of states here? Shouldn't we just reset existing states?

Reply:
This pattern exists pre-PR and is used to support multi-threading (which, per our decision, is currently not supported in segmented aggregation but is in regular aggregation). Each worker threads is associated with a unique index from 0 to `max_concurrency - 1`. The code here creates a dedicated state instance per thread.

Reply:
Do we need to reset it here those? Those shouldn't change between segments right?

Reply:
They state instances do need to be reset. They accumulate the aggregation per group. When the segment ends, the aggregation output is generated and the states are reset, ready for the next segment.

Reply:
I understand the state needs to be reset - my point is that we don't need to resize this because different segments should not change the size here

Reply:
Fixed.